Code/Core/Mem/SmallBlockAllocator.cpp

```diff
@@ -108,18 +108,18 @@ NO_INLINE void SmallBlockAllocator::InitBuckets()
 //------------------------------------------------------------------------------
 void * SmallBlockAllocator::Alloc( size_t size, size_t align )
 {
-    // Check we can support allocation of this size
-    if ( size > BUCKET_MAX_ALLOC_SIZE )
-    {
-        return nullptr; // Too big
-    }
-
     // Lazy initialization of buckets to support static allocations
     if ( s_BucketMemoryStart == nullptr )
     {
         InitBuckets();
     }
 
+    // Check we can support allocation of this size
+    if ( size > BUCKET_MAX_ALLOC_SIZE )
+    {
+        return nullptr; // Too big
+    }
```

Comment:
It makes sense to robustify this allocator in case behaviour changes such that it's possible to only perform large allocations and exit, but I think it's much better to do that by making the shutdown code that was crashing not crash (rather than forcing creation of buckets we don't need). Can you describe where it was crashing?

Reply:
Here is the crash stack:
 	ntdll.dll!RtlEnterCriticalSection()	Unknown
 	dummysandbox.exe!Mutex::Lock() Line 64	C++
>	dummysandbox.exe!SmallBlockAllocator::Free(void * ptr) Line 198	C++
 	dummysandbox.exe!Free(void * ptr) Line 82	C++
 	dummysandbox.exe!DefaultDeletor::Delete(void * ptr) Line 14	C++
 	dummysandbox.exe!AutoPtr<char,DefaultDeletor>::~AutoPtr<char,DefaultDeletor>() Line 25	C++
 	dummysandbox.exe!main(int argc, char * * argv) Line 93	C++
 	[External Code]	

Notice that the SmallBlockAllocator is being called but was never initialized, resulting in a crash when trying to acquire the bucket.m_Mutex.Lock() in SmallBlockAllocator::Free(). The lock object is invalid because SmallBlockAllocator Init() was never called. It was never called because all allocs in this process were too big and so returned from Alloc() early, before Init() could be called.

The fix I made here is to always Init(). If you think the Free() could be fixed up, instead, then that may work. I am unclear how, though. We could have a bool that says whether the SmallBlockAllocator was initialized or not, but wouldn't that need a lock around it?

Reply:
Ok, so the crash is in a new executable you have created, and not in some existing part of FASTBuild which explains why I have never seen it.

The SmallBlockAllocator already knows if it has been initilalized (s_BucketMemoryStart is nullptr). We could add an additional guard for that in Free. 

It's a shame however to add another branch to the free, but we can avoid that by changing the initial value s_BucketMemoryStart to a pointer outside of the valid user mode address space. This way, no large allocation can be within the memory range that would result in us calculating a valid pageIndex.

I've implemented that fix here: https://github.com/fastbuild/fastbuild/commit/4ac7570be393b9d11bad2dc953065f5bfc8e97dc

Reply:
Your fix works great. Thanks!